### PR TITLE
Update to tensorflow==2.7.0 - update requirements and build

### PR DESCRIPTION
**Pull request by lastmansleeping**
1. https://github.com/salesforce/ml4ir/pull/156
2. https://github.com/salesforce/ml4ir/pull/157
**3. https://github.com/salesforce/ml4ir/pull/158**
4. https://github.com/salesforce/ml4ir/pull/159

I was able to train both Ranking and EP models successfully with the following commands.
```
python ml4ir/applications/ranking/pipeline.py \
--data_dir ml4ir/applications/ranking/tests/data/tfrecord \
--feature_config ml4ir/applications/ranking/tests/data/configs/feature_config.yaml \
--run_id test \
--data_format tfrecord \
--loss_key softmax_cross_entropy \
--num_epochs 2 \
--execution_mode train_evaluate
```
and
```
python ml4ir/applications/classification/pipeline.py \
--data_dir ml4ir/applications/classification/tests/data/tfrecord \
--feature_config ml4ir/applications/classification/tests/data/configs/feature_config.yaml \
--model_config ml4ir/applications/classification/tests/data/configs/model_config.yaml \
--batch_size 32 \
--run_id test \
--data_format tfrecord \
--execution_mode train_evaluate \
--num_epochs 2
```

There are still a few pending items regardless and worth doing in follow up items before merging with master.

- [ x ] Fix ranking *validation* metrics. Test metrics are working fine.
- [ x ] Fix tests
- [ x ] [FOLLOW UP] `model.summary()` doesn't print the nested layers, this causes issues with temperature scaling and L1 which require accessing weights from nested layers.
- [ x ] [FOLLOW UP] Fix TF2.7 issue with handling duplicate metric names for old vs new metrics
- [ ] [FOLLOW UP]  Verify backward compatibility with `ml4ir-inference`
- [ ] [FOLLOW UP] Fix temperature scaling
- [ ] [FOLLOW UP] Fix transfer learning

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -148,6 +152,7 @@ jobs:
             python3 -m venv venv
             . venv/bin/activate
             cd python
+            pip install --upgrade pip
```
Comment (arvindsrikantan) [inline .circleci/config.yml:155]: Why was this necessary?

**Review by arvindsrikantan — APPROVED** [review]
Looks good!

File: .circleci/config.yml (hunk 1 above)
Comment (lastmansleeping) [reply .circleci/config.yml:155]: The base pip version on the circleci docker image wasn't able to install a package. I forget which one.

File: python/ml4ir/applications/classification/tests/test_classification_model.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def test_group_metrics_df(self):
         Test the dimensions of the grouped metrics
         """
         # Metrics cardinality and names
-        metrics = self.classification_pipeline.metrics_keys  # Metrics during training
+        metrics = ["loss"] + self.classification_pipeline.metrics_keys  # Metrics during training
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/classification/tests/test_classification_model.py:45]: Did you get a chance to train classification models with these changes @lastmansleeping ? Should we have a follow up item to ensure end-to-end flow works as expected?

File: python/ml4ir/applications/ranking/tests/test_losses.py [hunk 3]
```diff
@@ -28,38 +28,49 @@ def setUp(self):
...
+
+        y_pred = loss_op.final_activation_op({
+            "logits": self.logits,
+            "metadata": {
+                "mask": self.mask
+            }
+        })
         assert np.isclose(y_pred[0][0].numpy(), 0.54905695, atol=1e-5)
-        assert np.isclose(y_pred[2][4].numpy(), 0.64832306, atol=1e-5)
+        assert np.isclose(y_pred[2][4].numpy(), 0., atol=1e-5)
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/tests/test_losses.py:40]: Why did this change?

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_losses.py:40]: I think this should be expected to change really because we've modified underlying tensorflow, numpy versions which could be impacting the random seeds or even the implementations. This is my guess though.

File: python/ml4ir/applications/classification/tests/test_classification_model.py (hunk 2 above)
Comment (lastmansleeping) [reply python/ml4ir/applications/classification/tests/test_classification_model.py:45]: I have verified classification models too. You should be able to verify with
```
python ml4ir/applications/classification/pipeline.py \
--data_dir ml4ir/applications/classification/tests/data/tfrecord \
--feature_config ml4ir/applications/classification/tests/data/configs/feature_config.yaml \
--model_config ml4ir/applications/classification/tests/data/configs/model_config.yaml \
--batch_size 32 \
--run_id test \
--data_format tfrecord \
--execution_mode train_evaluate \
--num_epochs 2
```

Please note that the last step of this will fail with the default model_config because temperature scaling is not yet supported.

All other classification, ranking and base tests are passing though.

File: python/ml4ir/applications/ranking/tests/test_losses.py (hunk 3 above)
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_losses.py:40]: (Ignore my previous reply)

This is 0., and should be 0., because we mask the predictions too. We weren't doing this earlier. It wasn't a breaking issue though because it was being masked in the loss. I'm doing the masking in the activation now to make sure in the next version update, we can remove passing input features to the loss function like we discussed.

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_losses.py:40]: If you notice, the final loss computed is still the same in the next line.

**Review by arvindsrikantan — APPROVED** [review]
LGTM!
